src/main/java/com/basho/riak/client/core/RiakNode.java

```diff
@@ -691,82 +711,86 @@ private Channel doGetConnection() throws ConnectionFailedException
-                {
-                    context.init(null, tmf.getTrustManagers(), null);
-                }
-
-            }
-            catch (Exception ex)
-            {
-                c.close();
-                logger.error("Failure configuring SSL; {}:{} {}", remoteAddress, port, ex);
-                throw new ConnectionFailedException(ex);
-            }
+            setupTLSAndAuthenticate(c);
+        }
 
-            SSLEngine engine = context.createSSLEngine();
+        return c;
 
-            Set<String> protocols = new HashSet<String>(Arrays.asList(engine.getSupportedProtocols()));
+    }
 
-            if (protocols.contains("TLSv1.2"))
+    private void setupTLSAndAuthenticate(Channel c) throws ConnectionFailedException
+    {
+        SSLContext context;
+        try
+        {
+            context = SSLContext.getInstance("TLS");
+            TrustManagerFactory tmf =
+                TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+            tmf.init(trustStore);
+            if(keyStore!=null)
             {
-                engine.setEnabledProtocols(new String[] {"TLSv1.2"});
-                logger.debug("Using TLSv1.2");
+                KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+                kmf.init(keyStore, keyPassword==null?"".toCharArray():keyPassword.toCharArray());
+                context.init(kmf.getKeyManagers(), tmf.getTrustManagers(), null);
             }
-            else if (protocols.contains("TLSv1.1"))
+            else
             {
-                engine.setEnabledProtocols(new String[] {"TLSv1.1"});
-                logger.debug("Using TLSv1.1");
+                context.init(null, tmf.getTrustManagers(), null);
             }
 
-            engine.setUseClientMode(true);
-            RiakSecurityDecoder decoder = new RiakSecurityDecoder(engine, username, password);
-            c.pipeline().addFirst(decoder);
+        }
+        catch (Exception ex)
+        {
+            c.close();
+            logger.error("Failure configuring SSL; {}:{} {}", remoteAddress, port, ex);
+            throw new ConnectionFailedException(ex);
+        }
 
-            try
-            {
-                DefaultPromise<Void> promise = decoder.getPromise();
-                promise.await();
+        SSLEngine engine = context.createSSLEngine();
 
-                if (promise.isSuccess())
-                {
-                    logger.debug("Auth succeeded; {}:{}", remoteAddress, port);
-                }
-                else
-                {
-                    c.close();
-                    logger.error("Failure during Auth; {}:{} {}",remoteAddress, port, promise.cause());
-                    throw new ConnectionFailedException(promise.cause());
-                }
+        Set<String> protocols = new HashSet<String>(Arrays.asList(engine.getSupportedProtocols()));
 
+        if (protocols.contains("TLSv1.2"))
```

Comment:
I'm curious why not just pass the supported protocols as the enabled ones? TLS _should_ negotiate the highest supported version.